server/src/main/java/org/elasticsearch/search/query/QueryPhase.java

```diff
@@ -193,66 +184,59 @@ static void addCollectorsAndSearch(SearchContext searchContext) throws QueryPhas
-                    new SingleThreadCollectorManager(MultiCollector.wrap(terminateAfterCollector, collector)),
-                    REASON_SEARCH_TERMINATE_AFTER_COUNT,
-                    collector
-                );
+            Collector topDocsCollector = topDocsCollectorManager.newCollector();
+            Collector aggsCollector = null;
+            Weight postFilterWeight = null;
+            if (searchContext.aggregations() != null) {
+                aggsCollector = searchContext.aggregations().getAggsCollectorManager().newCollector();
             }
             if (searchContext.parsedPostFilter() != null) {
-                // add post filters before aggregations
-                // it will only be applied to top hits
-                final Weight filterWeight = searcher.createWeight(
+                postFilterWeight = searcher.createWeight(
                     searcher.rewrite(searchContext.parsedPostFilter().query()),
                     ScoreMode.COMPLETE_NO_SCORES,
                     1f
                 );
-                final Collector collector = collectorManager.newCollector();
-                collectorManager = wrapWithProfilerCollectorManagerIfNeeded(
-                    searchContext.getProfilers(),
-                    new SingleThreadCollectorManager(new FilteredCollector(collector, filterWeight)),
-                    REASON_SEARCH_POST_FILTER,
-                    collector
-                );
-            }
-            if (searchContext.aggregations() != null) {
-                final Collector collector = collectorManager.newCollector();
-                final Collector aggsCollector = searchContext.aggregations().getAggsCollectorManager().newCollector();
-                collectorManager = wrapWithProfilerCollectorManagerIfNeeded(
-                    searchContext.getProfilers(),
-                    new SingleThreadCollectorManager(MultiCollector.wrap(collector, aggsCollector)),
-                    REASON_SEARCH_MULTI,
-                    collector,
-                    aggsCollector
-                );
             }
-            if (searchContext.minimumScore() != null) {
-                final Collector collector = collectorManager.newCollector();
-                // apply the minimum score after multi collector so we filter aggs as well
-                collectorManager = wrapWithProfilerCollectorManagerIfNeeded(
-                    searchContext.getProfilers(),
-                    new SingleThreadCollectorManager(new MinimumScoreCollector(collector, searchContext.minimumScore())),
-                    REASON_SEARCH_MIN_SCORE,
-                    collector
-                );
+            QueryPhaseCollector queryPhaseCollector = new QueryPhaseCollector(
+                topDocsCollector,
+                postFilterWeight,
+                searchContext.terminateAfter(),
+                aggsCollector,
+                searchContext.minimumScore()
+            );
```

Comment:
Which logic are you referring to? How would it work to pass in the searcher? Right now it is a collector implementation which makes it an argument of the searcher#search call. The profiler wrapping makes it a different collector also, so I don't get how that would fit in it.

Reply:
I was thinking all the lines here could possibly become part of the collector's constructor in this case, but you're right it would be really strange with having to pass in the searcher.